pkgs/by-name/zl/zluda/package.nix

```diff
@@ -0,0 +1,80 @@
+{ lib, fetchFromGitHub, rocmPackages, python3, cargo, rustc, cmake, clang, zlib, libxml2, libedit, rustPlatform, stdenv }:
+
+rustPlatform.buildRustPackage rec {
+  pname = "zluda";
+  version = "3";
+
+  src = fetchFromGitHub {
+    owner = "vosen";
+    repo = "ZLUDA";
+    rev = "v${version}";
+    hash = "sha256-lykM18Ml1eeLMj/y6uPk34QOeh7Y59i1Y0Nr118Manw=";
+    fetchSubmodules = true;
+  };
+
+  buildInputs = [
+    rocmPackages.clr
+    rocmPackages.miopen
+    rocmPackages.rocm-smi
+    rocmPackages.rocsparse
+    rocmPackages.rocsolver
+    rocmPackages.rocblas
+    rocmPackages.hipblas
+    rocmPackages.rocm-cmake
+    rocmPackages.hipfft
+    zlib
+    libxml2
+    libedit
+  ];
+
+  nativeBuildInputs = [
+    python3
+    cargo
+    rustc
+    cmake
+    clang
+  ];
+
+  preConfigure = ''
+    # Comment out zluda_blaslt in Cargo.toml until hipBLASLt package is added: https://github.com/NixOS/nixpkgs/issues/197885#issuecomment-2046178008
+    sed -i '/zluda_blaslt/d' Cargo.toml
+    # disable test written for windows only: https://github.com/vosen/ZLUDA/blob/774f4bcb37c39f876caf80ae0d39420fa4bc1c8b/zluda_inject/tests/inject.rs#L55
+    rm zluda_inject/tests/inject.rs
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+    cargo xtask --release
+    runHook postBuild
+  '';
+  # xtask doesn't support passing --target, but nix hooks expect the folder structure from when it's set
+  env.CARGO_BUILD_TARGET = stdenv.hostPlatform.rust.cargoShortTarget;
+
+  cargoHash = "sha256-gZdLThmaeWVJXoeG7fuusfacgH2RNTHrqm8W0kqkqOY=";
+  cargoLock.lockFile = ./Cargo.lock;
+
+  # vergen panics if the .git directory isn't present
+  # Disable vergen and manually set env
+  postPatch = ''
+    substituteInPlace zluda/build.rs \
+      --replace-fail 'vergen(Config::default())' 'Some(())'
+    # ZLUDA repository missing Cargo.lock: vosen/ZLUDA#43
+    ln -s ${./Cargo.lock} Cargo.lock
+  '';
+  env.VERGEN_GIT_SHA = src.rev;
```

Comment:
Nit: could you move this section further up? You usually order sections in mkDerivation according to some sensible order; patch happens before build.

Reply:
@errnoh cf. the order of phases in the manual[^1], same as in the generic builder: https://github.com/NixOS/nixpkgs/blob/1049436671de4eab7ccd3682d33cdc087232c81d/pkgs/stdenv/generic/setup.sh#L1601-L1604

The customary "sensible order" of arguments to `mkDerivation` is by phases these arguments are most relevant to.


[^1]: https://nixos.org/manual/nixpkgs/unstable/#sec-stdenv-phases

Reply:
fixed in latest commit